queenbee/schema/function.py

```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
+            )
+
+        return v
+
+    @validator('command')
+    def validate_command_refs(cls, v, values):
+        """Validate referenced variables in the command"""
+
+        ref_var = qbvar.get_ref_variable(v)
+
+        # If inputs is not in values it has failed validation
+        # and we cannot check/validate output refs
+        if 'inputs' not in values:
+            return v
 
-        The validation checks are:
-            * Reference to workflow values is not allowed in a function.
-            * Referenced values are allowed in command and outputs. These values must
-              reference to function inputs.
-        """
-        input_names = {'parameters': [], 'artifacts': []}
         inputs = values.get('inputs')
-        if inputs:
-            irf = inputs.referenced_values
-            for key, value in irf.items():
-                if len(value) == 0:
-                    continue
-                raise ValueError(
-                    f'There is at least one referenced variable in inputs:\n'
-                    f'{key}: {value}.\n'
-                    f'Referenced values are not allowed in function inputs.'
-                )
-
-            if inputs.parameters:
-                input_names['parameters'] = [p.name for p in inputs.parameters]
-            if inputs.artifacts:
-                input_names['artifacts'] = [p.name for p in inputs.artifacts]
-        outputs = values.get('outputs')
-
-        if outputs:
-            # check output referenced values
-            orf = outputs.referenced_values
-            for rfv in orf.values():
-                if len(rfv) == 0:
-                    continue
-                for vv in rfv:
-                    for ovs in vv.values():
-                        for ov_refs in ovs.values():
-                            for ov in ov_refs:
-                                try:
-                                    names = input_names[ov.split('.')[1]]
-                                except KeyError:
-                                    names = []
-                                except Exception:
-                                    raise ValueError(
-                                        f'Invalid referenced value: {ov}'
-                                    )
-                            validate_function_ref_variables(ov, names)
-        # check command
-        command = ' '.join(values.get('command').split())
-        referenced_values = var_parser(command)
-        for ref_var in referenced_values:
-            try:
-                names = input_names[ref_var.split('.')[1]]
-            except KeyError:
-                names = []  # invalid input. validate function will throw the error
-            validate_function_ref_variables(ref_var, names)
-        return values
+
+        input_names = [param.name for param in inputs.parameters]
+
+        cls.validate_referenced_values(
+            input_names=input_names,
+            variables=ref_var
+        )
+
```

Comment:
`return v` is missing.

Reply:
Ah... that's why it was acting funny :upside_down_face: Good catch thanks!